src/Microsoft.AspNetCore.Mvc.Core/Formatters/TextOutputFormatter.cs

```diff
@@ -0,0 +1,258 @@
+            }
+
+            var request = context.HttpContext.Request;
+            var encoding = MatchAcceptCharacterEncoding(request.GetTypedHeaders().AcceptCharset);
+            if (encoding != null)
+            {
+                return encoding;
+            }
+
+            if (context.ContentType.HasValue)
+            {
+                var parsedContentType = new MediaType(context.ContentType);
+                var contentTypeCharset = parsedContentType.Charset;
+                if (contentTypeCharset.HasValue)
+                {
+                    for (var i = 0; i < SupportedEncodings.Count; i++)
+                    {
+                        var supportedEncoding = SupportedEncodings[i];
+                        if (contentTypeCharset.Equals(supportedEncoding.WebName, StringComparison.OrdinalIgnoreCase))
+                        {
+                            // This is supported.
+                            return SupportedEncodings[i];
+                        }
+                    }
+                }
+            }
+
+            return SupportedEncodings[0];
+        }
+
+        /// <inheritdoc />
+        public override Task WriteAsync(OutputFormatterWriteContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var selectedMediaType = context.ContentType;
+            if (!selectedMediaType.HasValue)
+            {
+                // If content type is not set then set it based on supported media types.
+                if (SupportedEncodings.Count > 0)
+                {
+                    selectedMediaType = new StringSegment(SupportedMediaTypes[0]);
+                }
+                else
+                {
+                    throw new InvalidOperationException(Resources.FormatOutputFormatterNoMediaType(GetType().FullName));
+                }
+            }
+
+            var selectedEncoding = SelectCharacterEncoding(context);
+            if (selectedEncoding != null)
+            {
+                // Override the content type value even if one already existed.
+                var mediaTypeWithCharset = GetMediaTypeWithCharset(selectedMediaType.Value, selectedEncoding);
+                selectedMediaType = new StringSegment(mediaTypeWithCharset);
+            }
+
+            context.ContentType = selectedMediaType;
+
+            WriteResponseHeaders(context);
+            return WriteResponseBodyAsync(context, selectedEncoding);
+        }
+
+        /// <inheritdoc />
+        public override async Task WriteResponseBodyAsync(OutputFormatterWriteContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var encoding = MediaType.GetEncoding(context.ContentType);
+            await WriteResponseBodyAsync(context, encoding);
```

Comment:
so basically this won't ever get called right? Why not seal it and throw an exception